Core: More Memory Cleanup & Fixes

StevenMiller123 opened the pull request:
The main change of this PR is completely removing our `PoolReserve` and `Reserve` functions in `Core::Memory`. Both of these functions were effectively re-implementations of `MapMemory` with minor changes, since all three functions are different uses of the `mmap` syscall.
Instead of constantly updating all three separately, it's easier to group all behavior in one function with additional checks where we need to behave slightly differently.

I've also added proper error handling in `posix_mmap`, added proper conditional rasterizer memory mapping logic based on protection values, and fixed a few other edge cases I've noticed in testing.

I'm opening this as a draft for now, as I need to make some more minor cleanup/style commits, and these changes need proper testing before any merges.

Commit:
Only perform GPU memory mapping when GPU can access it

This better aligns with hardware observations, and should also speed up unmaps and decommits, since they don't need to be compared with the GPU max address anymore.

Commit:
Reserve fixes

ReserveVirtualRange seems to follow the 0x200000000 base address like MemoryPoolReserve does.
Both also need checks in their flags Fixed path to ensure we're mapping in-bounds. If we're not in mapping to our address space, we'll end up reserving and returning the wrong address, which could lead to weird memory issues in games.

I'll need to test on real hardware to verify if such changes are appropriate.

Commit:
Better sceKernelMmap

Handles errors where we would previously throw exceptions. Also moves the file logic to MapFile, since that way all the possible errors are in one place.
Also fixes some function parameters to align with our current standards.

Commit:
Major refactor

MapDirectMemory, MapFlexibleMemory, ReserveVirtualRange, and MemoryPoolReserve all internally use mmap to perform their mappings. Naturally, this means that all functions have similar behaviors, and a lot of duplicate code.
This add necessary conditional behavior to MapMemory so MemoryPoolReserve and ReserveVirtualRange can use it, without disrupting the behavior of MapDirectMemory or MapFlexibleMemory calls.

Commit:
Accurate phys_addr for non-direct mappings

Commit:
Properly handle GPU access rights

Since my first commit restricts GPU mappings to memory areas with GPU access permissions, we also need to be updating the GPU mappings appropriately during Protect calls too.

Commit:
Update memory.cpp

Commit:
Update memory.h

Commit:
Update memory.cpp

Commit:
Update memory.cpp

Commit:
Update memory.cpp

Commit:
Revert "Update memory.cpp"

This reverts commit 2c55d014c0efbdfadee4121b01e1dcf5af60e63d.

Commit:
Coalesce dmem map

Aligns with hardware observations, hopefully shouldn't break anything since nothing should change hardware-wise when release dmem calls and unmap calls are performed?
Either that or Windows breaks because Windows, will need to test.

Commit:
Fix Mprotect on free memory

On real hardware, this just does nothing. If something did get protected, there's no way to query that information.
Therefore, it seems pretty safe to just behave like munmap and return size here.

Commit:
Implement posix_mprotect

Unity calls this
Also fixes the names of sceKernelMprotect and sceKernelMtypeprotect, though that's more of a style change and can be reverted if requested.

Commit:
Fix sceKernelSetVirtualRangeName

Partially addresses a "regression" introduced when I fixed up some asserts.
As noted in the code, this implementation is still slightly inaccurate, as handling this properly could cause regressions on Windows.

red-prig commented on src/core/memory.cpp:
```diff
@@ -449,8 +389,22 @@ int MemoryManager::MapFile(void** out_addr, VAddr virtual_addr, size_t size, Mem
+    if (prot >= MemoryProt::GpuRead) {
+        // PS4s only map to GPU memory when the protection includes GPU access.
+        // If the address to map to is too high, PS4s throw a page fault and crash.
+        ASSERT_MSG(IsValidGpuMapping(mapped_addr, size_aligned), "Invalid address for GPU mapping");
```
As far as I know, regular files cannot be made visible on the GPU, so I think there should be an unconditional Assert here.

Commit:
Unconditional assert in MapFile

red-prig commented on src/core/memory.cpp:
```diff
@@ -563,8 +563,9 @@ s64 MemoryManager::ProtectBytes(VAddr addr, VirtualMemoryArea vma_base, size_t s
-        LOG_ERROR(Kernel_Vmm, "Cannot change protection on free memory region");
-        return ORBIS_KERNEL_ERROR_EINVAL;
+        // On PS4, protecting freed memory does nothing.
+        LOG_WARNING(Kernel_Vmm, "Attempting to protect free memory");
```
Free regions can simply be skipped in mprotect, this is expected behavior

StevenMiller123 replied:
That's what this change does. By returning size here, the loop we run will skip the free VMA and move to the next one. https://github.com/StevenMiller123/shadPS4/blob/e2215f0f4f47d5b5b1dfe3a414702ea19630d3f4/src/core/memory.cpp#L625
Will remove the log though, since it's probably not important to log this case.

Commit:
Remove protect warning

This is expected behavior, shouldn't need any logging.

red-prig commented on src/core/memory.cpp:
```diff
@@ -449,8 +389,19 @@ int MemoryManager::MapFile(void** out_addr, VAddr virtual_addr, size_t size, Mem
+
+    const auto handle = file->f.GetFileMapping();
+
+    impl.MapFile(mapped_addr, size_aligned, phys_addr, std::bit_cast<u32>(prot), handle);
```
In case of file mappings there is a complicated case when the file size is not a multiple of the page size, in this case the upper part of the mapped page (the virtual address is larger than the file) is always equally available for reading/writing and is filled with zeros by default, the problem arises firstly because Windows requires specifying an unaligned file size otherwise an error (Does Linux allow an aligned size larger than the file? Correct me) and also there is a problem with the fact that the host page multiple is 4KB and the guest page multiple is 16KB, so additional code will be needed to add private pages for alignment to the entire guest page (for example, the file size is 3KB, it turns out one virtual page of 4KB for the host, so at the end you will need to add another 3 4KB private pages)

StevenMiller123 replied:
I'm not too familiar with our address space code yet, which is why my changes to our file mapping code are very minimal in this PR. All I did was move around a couple lines (and made that GPU mapping change I've already reverted).

From what I can tell, our implementation on both Windows and Linux allow for the aligned up size we've been using here. On Windows we're using MapViewOfFile3 for most cases, and VirtualAlloc2 + ReadFile for read-only file mappings. On Linux we use the host's mmap. So far, I don't think I've seen any games crop up that fail on these calls, though we've got asserts in place to identify any errors.
As for reading the upper part of the page, apparently our current Linux implementation will throw a SIGBUS exception looking at their mmap docs. I'm not sure how our Windows implementation behaves here.

I'll try taking a look at this issue at some point, but I'm not sure if I'll fix that in this PR or not, since I'll need to learn how all these functions work first.

PirkyP replied:
Sorry for the delay. I ran the test and it's a different error now. The boot screen displayed for a few seconds before it crashed. Let me know if I need to retest it.

`[Debug] <Critical> vector_memory.cpp:182 operator(): Assertion Failed!`
`Non immediate offset not supported`
[shad_log.txt](https://github.com/user-attachments/files/20474790/shad_log.txt)

Commit:
Respect alignment

Forgot to properly do this when updating ReserveVirtualRange and MemoryPoolReserve

Commit:
Minor tidy-up

No functional difference, but looks better.
